File: AED/countingSort.c
/*
 * Programa : countinSort.c
 * Objetivo : implementação de um counting sort
 * Autor : Augusto dos Santos
 * Data : 08/09/2022
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "countingSort.h"

int
getMaior(int * vetor, int tamVetor)
{
    int maior = 0;
    for(int i = 0; i < tamVetor; i++)
        if (vetor[i] > maior)
            maior = vetor[i];
    return maior;
}

//retorna o número naquele determinado dígito
int
getDigit(int numero, int digit)
{
    return numero / (int) pow (10,digit) % 10;
}

//devolve o maior elemento do vetor em um digito específico
int
getMaiorDigit(int * vetor, int tamVetor, int digit)
{
    int maior = 0;
    int vetorAtual;
    for(int i = 0; i < tamVetor; i++)
    {   
        vetorAtual = getDigit(vetor[i], digit);
        if (vetorAtual > maior)
            maior = vetorAtual;
    }
    return maior;
}

int *
getFreq(int * vetor, int tamVetor)
{
    int * vetorFreq = (int *) calloc (sizeof(int), getMaior(vetor, tamVetor));
    for(int i = 0; i < tamVetor; i++)
        vetorFreq[vetor[i]]++;
    return vetorFreq;
}

//retorna a frequencia de cada valor de um determinado dígito
int *
getFreqDigit(int * vetor, int tamVetor, int digit)
{
    int * vetorFreq = (int *) calloc (sizeof(int), getMaiorDigit(vetor, tamVetor, digit));
    for(int i = 0; i < tamVetor; i++)
        vetorFreq[getDigit(vetor[i], digit)]++;
    return vetorFreq;
}

int *
getFreqAcumulada(int * vetorFreq, int tamVetor)
{
    int * vetorFreqAcumuladas = (int *) calloc (sizeof(int), tamVetor);

    vetorFreqAcumuladas[0] = vetorFreq[0];
    
    for(int i = 1; i < tamVetor; i++)
        vetorFreqAcumuladas[i] += vetorFreqAcumuladas[i - 1] + vetorFreq[i - 1];
    
    return vetorFreqAcumuladas;
}

int *
countingSort(int * vetor, int tamVetor)
{
    int * vetorOrdenado = (int *) malloc (sizeof(int) * tamVetor);
    int * vetFreq = getFreq(vetor, tamVetor);
    int * vetFreqAcumuladas = getFreqAcumulada(vetFreq, getMaior(vetor, tamVetor) + 1);

    for(int i = tamVetor -1; i >= 0; i--)
        vetorOrdenado[--vetFreqAcumuladas[vetor[i]]] = vetor[i];
    
    for(int i = 0; i < tamVetor; i++)
        vetor[i] = vetorOrdenado[i];

    free(vetFreq);
    free(vetFreqAcumuladas);
    free(vetorOrdenado);
}

int *
countingSortDigit(int * vetor, int tamVetor, int digit)
{
    int * vetorOrdenado = (int *) malloc (sizeof(int) * tamVetor);
    int * vetFreq = getFreqDigit(vetor, tamVetor, digit);
    int * vetFreqAcumuladas = getFreqAcumulada(vetFreq, getMaiorDigit(vetor, tamVetor, digit));

    for(int i = tamVetor -1; i >= 0; i--)
        vetorOrdenado[--vetFreqAcumuladas[getDigit(vetor[i], digit)]] = vetor[i];
    
    for(int i = 0; i < tamVetor; i++)           //copiando o vetor ordenado para o vetor original
        vetor[i] = vetorOrdenado[i];

    free(vetFreq);
    free(vetFreqAcumuladas);
    free(vetorOrdenado);
}

int
main()
{
    int * vetor = (int *) malloc (sizeof(int) * 10);
    vetor[0] = 90;
    vetor[1] = 81;
    vetor[2] = 72;
    vetor[3] = 63;
    vetor[4] = 54;
    vetor[5] = 45;
    vetor[6] = 36;
    vetor[7] = 27;
    vetor[8] = 18;
    vetor[9] = 9;

    countingSort(vetor, 10);
    for(int i = 0; i < 10; i++)
        printf("%d ", vetor[i]);


    return 0;
}

File: AED/countingSort.py
'''
Programa : countingSort.py
Objetivo : implementação de um algoritmo de counting sort
Autor : Augusto dos Santos Gomes Vaz
Data : 08/09/2022
'''

'''
get Digito
Esta função retorna o dígito especificado
'''
from cmath import log
from itertools import count


def getDigito(numero, digito):
    return int(numero / 10 ** (digito - 1)) % 10

'''
Max digitos do vetor
Retorna a quantidade máxima de dígitos que tem o vetor
'''
#def maiorDigitos(vetor):
#    return int(log(max(vetor)))

#vetor = [100, 200]
#print(maiorDigitos(vetor))
    
'''
Max vetor digito
Retorno o maior elemento de um determinado dígito em um vetor de inteiros
'''
def maxDigito(vetor, digito):
    # Determinando o numero de casas
    var = 0
'''
get Vetor Frequencias
Esta função cria um vetor que contém a quantidade de vezes que cada número do vetor aparece;
O vetor tem tamanho igual ao maior valor presente no vetor;
'''
def getFreq(vetor):
    vetFreq = [0] * (max(vetor) + 1)

    for i in vetor:
        vetFreq[i] += 1
    return vetFreq

'''
get Vetor Frequencias Digito
Esta função cria um vetor que contém a quantidade de vezes que cada número do vetor aparece;
O vetor tem tamanho igual ao maior valor presente no vetor;
'''
def getFreqDig(vetor, digito):
    vetFreq = [0] * (max(vetor) + 1)

    for i in vetor:
        vetFreq[i] += 1
    return vetFreq

'''
Get Frequencias Acumuladas
Retorna a quantidade de vezes que aparecem números antes do número atual
'''
def getFreqAcumulada(vetFreq):
    vetFreqAcumulada = [0] * (len(vetFreq))
    for i in range(1, len(vetFreq)):
        vetFreqAcumulada[i] = vetFreqAcumulada[i - 1] + vetFreq[i]
        
    return vetFreqAcumulada

'''
couting sort
retorna o vetor ordenado usando o método counting sort estável;
Precisa acessar o vetor desordenado para funcionar
'''
def countingSort(vetor):
    vetOrdenado = [0] * len(vetor)
    vetFreqAcumulada = getFreqAcumulada(getFreq(vetor))
    for i in reversed(vetor):
        vetFreqAcumulada[i] -= 1
        vetOrdenado[vetFreqAcumulada[i]] = i

    return vetOrdenado

'''

'''
def countingSortDigito(vetor, digito):
    vetOrdenado = [0] * len(vetor)
    vetFreqAcumulada = getFreqAcumulada(getFreq(vetor))
    
    for i in reversed(vetor):
        vetFreqAcumulada[i] -= 1
        vetOrdenado[vetFreqAcumulada[i]] = i

    return vetOrdenado


# main
vetor = [3, 0, 1, 3]
vetor = countingSort(vetor)
print(vetor)